File: agent-sdk/typescript/examples/llm.ts
// llm.ts — a tiny streaming chat client that targets EITHER a local Ollama or
// OpenRouter, chosen automatically from the environment.
//
// This is the "reusable base" behind `04-combined.ts`: it reduces both backends
// to a single chat shape, so the agent — and any future tool-calling — looks the
// same regardless of provider. Keep it small and dependency-free (just `fetch`).
//
//   OPENROUTER_API_KEY set?  → OpenRouter   (OPENROUTER_MODEL, default openai/gpt-4o-mini)
//   otherwise                → local Ollama (OLLAMA_MODEL, default llama3.2; OLLAMA_URL)

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
}

export interface LlmClient {
  /** Human-readable "backend/model", handy for logging. */
  readonly label: string;
  /** Stream the assistant's reply to `messages`, yielding text as it arrives. */
  chatStream(messages: ChatMessage[]): AsyncGenerator<string>;
}

/** Pick a backend from the environment: OpenRouter if a key is present, else Ollama. */
export function createLlmClient(): LlmClient {
  const apiKey = process.env["OPENROUTER_API_KEY"];
  return apiKey ? openRouter(apiKey) : ollama();
}

// --- Ollama (local) ---------------------------------------------------------
function ollama(): LlmClient {
  const url = process.env["OLLAMA_URL"] ?? "http://localhost:11434";
  const model = process.env["OLLAMA_MODEL"] ?? "llama3.2";
  return {
    label: `ollama/${model}`,
    async *chatStream(messages) {
      const res = await fetch(`${url}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ model, messages, stream: true }),
      });
      if (!res.ok || res.body === null) {
        throw new Error(`Ollama request failed: ${res.status} ${res.statusText}`);
      }
      // /api/chat returns newline-delimited JSON, each line `{message:{content}}`.
      const decoder = new TextDecoder();
      let buffer = "";
      for await (const bytes of res.body) {
        buffer += decoder.decode(bytes as Uint8Array, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";
        for (const line of lines) {
          if (line.trim() === "") continue;
          // Tolerate a rare malformed line instead of throwing out of the
          // generator (same guard as the OpenRouter branch and 05-tools.ts).
          try {
            const token =
              (JSON.parse(line) as { message?: { content?: string } }).message?.content ?? "";
            if (token) yield token;
          } catch {
            /* skip malformed line */
          }
        }
      }
    },
  };
}

// --- OpenRouter (hosted, OpenAI-compatible) ---------------------------------
function openRouter(apiKey: string): LlmClient {
  const model = process.env["OPENROUTER_MODEL"] ?? "openai/gpt-4o-mini";
  return {
    label: `openrouter/${model}`,
    async *chatStream(messages) {
      const res = await fetch("https://openrouter.ai/api/v1/chat/completions", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ model, messages, stream: true }),
      });
      if (!res.ok || res.body === null) {
        throw new Error(`OpenRouter request failed: ${res.status} ${res.statusText}`);
      }
      // OpenAI SSE: `data: {json}` lines (+ keep-alive comments), then `data: [DONE]`.
      const decoder = new TextDecoder();
      let buffer = "";
      for await (const bytes of res.body) {
        buffer += decoder.decode(bytes as Uint8Array, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";
        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed.startsWith("data:")) continue;
          const data = trimmed.slice(5).trim();
          if (data === "" || data === "[DONE]") continue;
          try {
            const token =
              (JSON.parse(data) as { choices?: { delta?: { content?: string } }[] }).choices?.[0]
                ?.delta?.content ?? "";
            if (token) yield token;
          } catch {
            /* ignore the rare non-JSON keep-alive line */
          }
        }
      }
    },
  };
}

File: agent-sdk/python/examples/llm.py
# llm.py — a tiny streaming chat client that targets EITHER a local Ollama or
# OpenRouter, chosen automatically from the environment. Python mirror of
# agent-sdk/typescript/examples/llm.ts.
#
# This is the "reusable base" behind 04-combined.py: it reduces both backends to
# one chat shape, so the agent — and any future tool-calling — looks the same
# regardless of provider. Keep it small and dependency-light (just httpx).
#
#   OPENROUTER_API_KEY set?  → OpenRouter (OPENROUTER_MODEL, default openai/gpt-4o-mini)
#   otherwise                → local Ollama (OLLAMA_MODEL, default llama3.2; OLLAMA_URL)

from __future__ import annotations

import json
import os
from collections.abc import AsyncGenerator

import httpx

ChatMessage = dict[str, str]  # {"role": ..., "content": ...}


class OllamaClient:
    def __init__(self) -> None:
        self.url = os.environ.get("OLLAMA_URL", "http://localhost:11434")
        self.model = os.environ.get("OLLAMA_MODEL", "llama3.2")
        self.label = f"ollama/{self.model}"

    async def chat_stream(self, messages: list[ChatMessage]) -> AsyncGenerator[str, None]:
        # /api/chat returns newline-delimited JSON, each line `{message: {content}}`.
        async with (
            httpx.AsyncClient(timeout=None) as client,
            client.stream(
                "POST",
                f"{self.url}/api/chat",
                json={"model": self.model, "messages": messages, "stream": True},
            ) as resp,
        ):
            resp.raise_for_status()
            async for line in resp.aiter_lines():
                if not line.strip():
                    continue
                # A rare malformed line shouldn't crash the stream — skip it.
                try:
                    token = (json.loads(line).get("message") or {}).get("content", "")
                except json.JSONDecodeError:
                    continue
                if token:
                    yield token


class OpenRouterClient:
    def __init__(self, api_key: str) -> None:
        self.api_key = api_key
        self.model = os.environ.get("OPENROUTER_MODEL", "openai/gpt-4o-mini")
        self.label = f"openrouter/{self.model}"

    async def chat_stream(self, messages: list[ChatMessage]) -> AsyncGenerator[str, None]:
        # OpenAI SSE: `data: {json}` lines (+ keep-alive comments), then `data: [DONE]`.
        headers = {"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"}
        async with (
            httpx.AsyncClient(timeout=None) as client,
            client.stream(
                "POST",
                "https://openrouter.ai/api/v1/chat/completions",
                headers=headers,
                json={"model": self.model, "messages": messages, "stream": True},
            ) as resp,
        ):
            resp.raise_for_status()
            async for raw in resp.aiter_lines():
                line = raw.strip()
                if not line.startswith("data:"):
                    continue
                data = line[len("data:") :].strip()
                if data in ("", "[DONE]"):
                    continue
                try:
                    token = json.loads(data)["choices"][0]["delta"].get("content")
                except (json.JSONDecodeError, KeyError, IndexError):
                    continue
                if token:
                    yield token


LlmClient = OllamaClient | OpenRouterClient


def create_llm_client() -> LlmClient:
    """Pick a backend from the environment: OpenRouter if a key is present, else Ollama."""
    api_key = os.environ.get("OPENROUTER_API_KEY")
    return OpenRouterClient(api_key) if api_key else OllamaClient()
